### ISSUE TITLE
research: formalize SRP as equivalence classes of methods by Reach-profile (deep, multi-theory)

### ISSUE THREAD

**Issue by mikeshogin** [mshogin/archlint#154]
## Goal

Formalize Single Responsibility Principle (SRP) mathematically. "A class has one reason to change" -> precise graph-theoretic definition.

## Theoretical framework

For each method `m` of type `T`, define:
```
Reach(m) = FieldAccess(m) ∪ ExternalCalls(m)
```
where:
- FieldAccess(m) = set of fields of T that m reads/writes (structural, from expanded graph)
- ExternalCalls(m) = set of external packages/types that m calls (behavioral, from callgraph)

Responsibility count of T = number of equivalence classes of methods under relation:
- m_i ~ m_j iff Reach(m_i) ∩ Reach(m_j) ≠ ∅
- OR m_i calls m_j (or vice versa)
- OR both Reach(m_i) and Reach(m_j) have empty external part

This uses **transitive clustering** (connected components), not pairwise thresholds — avoids false positives when methods are linked through intermediates.

## Cross-validation through multiple theories

The same fact (number of responsibilities) should be verifiable through independent mathematical lenses. If 3+ theories agree, the signal is robust. Research:

1. **Set theory**: Reach intersection + connected components (above)
2. **Graph theory**: Community detection (Louvain, label propagation) on method-interaction graph. Modularity Q score.
3. **Information theory**: Shannon entropy of method reach distribution. I(method; field) mutual information. Information bottleneck for responsibility compression.
4. **Spectral theory**: Laplacian eigenvalues of method similarity matrix. Fiedler vector for bipartition. Algebraic connectivity.
5. **Topology**: Čech complex on reach-profile vectors. Persistent homology. Betti numbers as dimension of responsibility space.

## Research deliverables

- [ ] Formal definition of SRP with precise notation
- [ ] 5 theoretical treatments (one per theory above), each with:
  - Mathematical formulation
  - Algorithm (pseudocode)
  - Academic references (Hitz & Montazeri, Chidamber-Kemerer, Martin, Newman modularity, von Luxburg spectral, Carlsson TDA)
  - Strengths/weaknesses
- [ ] Golden dataset: 20-30 hand-labeled types (data class / focused service / god object / mixed) from real projects
- [ ] Cross-validation matrix: which theories agree on which cases, precision/recall per theory
- [ ] Critiques anticipated from domain experts + rebuttals
- [ ] Relationship to existing metrics (LCOM1-5, CK metrics, MOOD)
- [ ] Proposed differences: why this is not LCOM4 rebadged

## Article-grade requirements

- Article title target: "SOLID as graph invariants: formalizing Single Responsibility Principle"
- Audience: senior architects, professional conference critics
- Every claim backed by either academic reference or empirical measurement
- Mathematical rigor: proper theorem-proof structure where applicable

## Acceptance criteria

- [ ] Research document `docs/solid-srp-formalization.md` (~40-50 pages material level)
- [ ] 5 theoretical sections complete
- [ ] Golden dataset with annotations
- [ ] Cross-validation matrix populated
- [ ] Bibliography (academic papers)
- [ ] Open questions and future work section

## Dependencies

- #153 (extended graph with field nodes) must be done first
